llm_toolkit/models.py

```diff
@@ -607,24 +616,82 @@ class GeminiV1D5(GeminiModel):
 class GeminiV1D5Chat(GeminiV1D5):
   """Gemini 1.5 for chat session."""
   name = 'vertex_ai_gemini-1-5-chat'
+  _vertex_ai_model = 'gemini-1.5-pro-002'
+
+  # Avoids sending large prompt as requested by Gemini team.
+  MAX_INPUT_TOKEN: int = 128000  # max 2000000
 
   def get_chat_client(self, model: GenerativeModel) -> Any:
     return model.start_chat(response_validation=False)
 
-  @retryable(exceptions=[
-      GoogleAPICallError,
-      InvalidArgument,
-  ],
-             other_exceptions={ResourceExhausted: 100})
+  @retryable(
+      exceptions=[
+          GoogleAPICallError,
+          InvalidArgument,
+          ValueError,  # TODO(dongge): Handle RECITATION specifically.
+          IndexError,  # A known error from vertexai.
+      ],
+      other_exceptions={ResourceExhausted: 100})
   def _do_generate(self, client: ChatSession, prompt: str,
                    config: dict[str, Any]) -> Any:
     """Generates chat response."""
     logger.info('%s generating response with config: %s', self.name, config)
-    return client.send_message(
-        prompt,
-        stream=False,
-        generation_config=config,
-        safety_settings=self.safety_config).text  # type: ignore
+    try:
+      return client.send_message(
+          prompt,
+          stream=False,
+          generation_config=config,
+          safety_settings=self.safety_config).text  # type: ignore
+    except Exception as e:
+      logger.error('%s failed to generated response: %s; Config: %s', e,
+                   self.name, config)
+      return ''
+
+  def truncate_prompt(self,
+                      raw_prompt_text: Any,
+                      extra_text: Any = None) -> Any:
+    """Truncates the prompt text to fit in MAX_INPUT_TOKEN."""
+    raw_prompt_text = raw_prompt_text or ' '
+
+    # A rough estimation in case count_tokens failed on very large prompt.
+    # Gemini suggest 4 char per token, using 3 here to be safer. Normally,
+    # prompts should not be this large anyway.
+    if len(raw_prompt_text) > 3 * self.MAX_INPUT_TOKEN:
+      raw_prompt_text = raw_prompt_text[-3 * self.MAX_INPUT_TOKEN:]
+
+    try:
+      original_token_count = self.get_model().count_tokens(
+          raw_prompt_text).total_tokens
+      if extra_text:
+        extra_token_count = self.get_model().count_tokens(
+            extra_text).total_tokens
+      else:
+        extra_token_count = 0
+      token_count = original_token_count
+    except Exception as e:
```

Comment:
we should avoid this error handling if we can. what exception do we see here? If we absolutely have to, let's have this all be handled inside `estimate_token_num` ?

Reply:
I've replaced this code with `estimate_token_num()` for simplicity.
The `count_tokens()` API has some drawbacks, (e.g., unavailable in `us-west2`) and we do not need the exact token number in this case, particularly given we have largely reduce max context token size.